Add options for client and server tracing to skip client errs

taywrobel opened the pull request:
This change introduces two configuration options, one for the
server hooks, and one for the client, which configure whether
client-class errors (4xx) are recorded as "error: true" in the
span.

Often times, client errors are expected behavior from the
perspective of the service, and having them report as errors to
trace collectors can obfuscate real service issues.

This is implemented using a functional options paradigm with
var args, which maintains backwards compatibility. Current users
will see no change to behavior or method signatures when updating.

taywrobel committed:
Add options for client and server tracing to skip client errs

This change introduces two configuration options, one for the
server hooks, and one for the client, which configure whether
client-class errors (4xx) are recorded as "error: true" in the
span.

Often times, client errors are expected behavior from the
perspective of the service, and having them report as errors to
trace collectors can obfuscate real service issues.

This is implemented using a functional options paradigm with
var args, which maintains backwards compatibility. Current users
will see no change to behavior or method signatures when updating.

iheanyi commented on trace_http_client.go:
```diff
@@ -20,18 +20,36 @@ type HTTPClient interface {
 type TraceHTTPClient struct {
 	client HTTPClient
 	tracer opentracing.Tracer
+	withUserErr bool
```
Two things:

1) I think it'd be worth it to make a new `ClientOptions` struct that encapsulates all of the configurable options for the `TraceHTTPClient` so we don't populate the main `TraceHTTPClient` struct with configuration options.  Then we can add `opts ClientOptions` to `TraceHTTPClient` and make something like:


```go
defaultOpts := &ClientOptions{
    includeClientErrors: true,
}

for _, opt := range opts {
    opt(defaultOpts)
}

return &TraceHTTPClient{
    client: client,
    tracer: tracer,
    opts: defaultOpts,
}
```

This will also leave more room for more extensibility in the future.

Second thing is, how do you feel about renaming it from `withUserErr` to `includeClientErrors`? I personally think it's a bit more semantic. Thoughts?

taywrobel replied:
Done!

And agree with `includeClientErrors`.  It's a tricky one to name since it's flagging whether to set errors as errors, just in a different context.

iheanyi commented on trace_http_client.go:
```diff
@@ -20,18 +20,36 @@ type HTTPClient interface {
+
+// WithClientUserErr, if set, will report client errors (4xx) as errors in the span.
+// If not set, only 5xx status will be reported as erroneous to the tracer.
+func WithClientUserErr(withUserError bool) ClientOpt {
```
Keeping this inline with my earlier comment, how do you feel about renaming this to `IncludeClientErrors(includeClientErrors bool) ClientOpt`? I know that functional options normally use the `With` prefix, but I figured `Include` may be a little more clear.

iheanyi commented on trace_http_client_test.go:
```diff
@@ -20,6 +20,7 @@ func TestTraceHTTPClient(t *testing.T) {
 		desc         string
 		errExpected  bool
 		service      twirptest.Haberdasher
+		clientOpts   []TraceClientOpt
```
Did you mean for this to be `ClientOpt`?
```suggestion
		clientOpts   []ClientOpt
```

iheanyi commented on tracing.go:
```diff
@@ -22,17 +22,31 @@ type tracingInfoKey struct{}
 	return hooks
 }
 
+type HookOpt func(h *twirp.ServerHooks)
```
Hm, rather than naming this `HookOpt`, how do you feel about naming this `TraceServerOpt`? 

Also, it feels weird using functional options to override the entire `*twirp.ServerHooks` methods. I'm wondering if we can make a wrapper struct like this:

```go
type TracingServerHooks struct {
    opts *TraceServerOptions
}

type TraceServerOptions struct {
     IncludeClientErrors bool
     Tracer opentracing.Tracer
}

func IncludeClientErrors(includeClientErrors bool) TraceServerOpt {
        return func(opts *TraceServerOptions) {
               opts.IncludeClientErrors = includeClientErrors
        }
}

func NewOpenTracingHooks(tracer ot.Tracer, opts ...TraceServerOpt) *twirp.ServerHooks {
        defaultOpts := &TraceServerOptions{
              IncludeClientErrors: true,
              Tracer: tracer,
        }

        for _, opt := range opts {
             opt(defaultOpts)
        }

        traceHooks := &TraceServerHooks{
                    opts: defaultOpts,
         }
    
         return traceHooks.TwirpHooks()
}

func (t *TraceServerHooks) TwirpHooks() *twirp.ServerHooks {
        return &twirp.ServerHooks{
               RequestReceived: t.startTraceSpan(t.opts.tracer), // optionally this tracer can actually be left out since we have direct access to it
               RequestRouted: t.handleRequestRouted,
               ResponseSent: t.finishTrace,
               Error: t.handleError,
        }
}

func (t *TraceServerHooks) handleError(ctx context.Context, err twirp.Error) context.Context {
    // Reimplement `handleError` logic here.
    // ...
    if t.opts.IncludeClientErrors || statusCode >= 500 {
         span.SetTag("error", true)
    }
}
```
Or alternatively, we can just make the `opts` it's own struct and pass that into every single function that we have defined. What do you think?

taywrobel replied:
Done!  I didn't put the tracer in the opts since it's a hard requirement and options generally are optional, but everything else worked out smoothly.

iheanyi commented on tracing.go:
```diff
@@ -22,17 +22,31 @@ type tracingInfoKey struct{}
+
+// WithUserErr, if set, will report client errors (4xx) as errors in the server span.
+// If not set, only 5xx status will be reported as erroneous.
+func WithUserErr(withUserError bool) HookOpt {
```
Similar to the client, how do you feel about changing this to be `IncludeClientErrors` instead?

taywrobel replied:
Done!  Tho, one thing I'm not the biggest fan of is that since these are in the same package they share a namespace, so both options can't be called `IncludeClientErrors`.  I ended up making the client one `IncludeUserErrors`, since from its perspective it's making the "error" but it's a user-class code, and the server one `IncludeClientErrors`, since its client-side for the server, but I don't think that's perfect/is still confusing.

If you have better recommendations on naming LMK.

iheanyi replied:
Thanks for changing this up! So, looking at the [grpc implementation](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/tracing/opentracing/client_interceptors.go) (which is what inspired this package), they actually share the same option. Should we consolidate these options on our end as well? 🤔 Maybe into `TraceOption`?

taywrobel replied:
Done and done!

iheanyi reviewed (changes requested):
Hey there, thanks for the pull request. Overall the logic looks sound, but I'd like for us to encapsulate as much of the functional options in a dedicated struct as possible. On the client side, this is just wrapping all the configurable values in a `ClientOptions` struct. On the server, I was thinking about adding a wrapper struct, but we may be able to get away with just using a `TraceServerOptions` struct for encapsulating the options and configuring them and passing this into every single server hook method. Let me know your thoughts and if you want help implementing some of these changes, happy to get them in before my day is over.

taywrobel committed:
Pull client options into a dedicated struct within the TraceHTTPClient

taywrobel committed:
Move server side trace opts into dedicated struct

In response to PR feedback, moves the server side configuraiton
into a dedicated struct, and introduces an adapter hook type to
configure the behavior of the hooks correctly based on the the
configuration options.

taywrobel committed:
Consolidate options structs

iheanyi reviewed (approved):
LGTM, thanks for the quick turnaround and being so receptive!
